include/crab/domains/object_domain.hpp

```diff
@@ -0,0 +1,3312 @@
+  void region_cast(const variable_t &src_rgn,
+                   const variable_t &dst_rgn) override {
+    // A region_cast is used to cast unknown regions to typed regions,
+    // or viceversa.
+    crab::CrabStats::count(domain_name() + ".count.region_cast");
+    crab::ScopedCrabStats __st__(domain_name() + ".region_cast");
+
+    CRAB_LOG("object", crab::outs()
+                           << "After region_cast(" << src_rgn << ":"
+                           << src_rgn.get_type() << "," << dst_rgn << ":"
+                           << dst_rgn.get_type() << ")=" << *this << "\n";);
+  }
+
+  // Remove a reference ref within region reg
+  void ref_free(const variable_t &reg, const variable_t &ref) override {}
+
+  // This default implementation is expensive because it will call the
+  // join.
+  // DEFAULT_SELECT_REF(object_domain_t)
+  void select_ref(const variable_t &lhs_ref, const variable_t &lhs_rgn,
+                  const variable_t &cond, const variable_or_constant_t &ref1,
+                  const boost::optional<variable_t> &rgn1,
+                  const variable_or_constant_t &ref2,
+                  const boost::optional<variable_t> &rgn2) override {
+    crab::CrabStats::count(domain_name() + ".count.select_ref");
+    crab::ScopedCrabStats __st__(domain_name() + ".select_ref");
+    if (!is_bottom()) {
+      // REDUCTION: perform reduction
+      perform_reduction();
+
+      ghost_variables_t lhs_ref_gvars = get_or_insert_gvars(lhs_ref);
+      base_dom_variable_t lhs_b_rgn =
+          is_unknown_region(lhs_rgn) ? lhs_rgn
+                                     : get_or_insert_gvars(lhs_rgn).get_var();
+      ghost_variables_t cond_gvars = get_or_insert_gvars(cond);
+      base_dom_variable_or_constant_t b_ref1 =
+          rename_variable_or_constant(ref1);
+      base_dom_variable_or_constant_t b_ref2 =
+          rename_variable_or_constant(ref2);
+      boost::optional<base_dom_variable_t> b_rgn1 =
+          rename_variable_optional(rgn1);
+      boost::optional<base_dom_variable_t> b_rgn2 =
+          rename_variable_optional(rgn2);
+
+      m_base_dom.select_ref(lhs_ref_gvars.get_var(), lhs_b_rgn, cond, b_ref1,
+                            b_rgn1, b_ref2, b_rgn2);
+      // ref_var is a reference, assign a fresh symbol into address dom
+      // create a fresh symbol to represent its base address
+      m_addrs_dom.set(get_or_insert_base_addr(lhs_ref),
+                      object_domain_impl::make_fresh_symbol(m_addrs_dom));
+    }
+  }
+
+  /**************************** Numerical operations *************************/
+  // x := y op z
+  void apply(arith_operation_t op, const variable_t &x, const variable_t &y,
+             const variable_t &z) override {
+
+    crab::CrabStats::count(domain_name() + ".count.apply");
+    crab::ScopedCrabStats __st__(domain_name() + ".apply");
+
+    if (!is_bottom()) {
+      // REDUCTION: perform reduction
+      perform_reduction();
```

Comment:
If reducing is a problem, add a flag about what has not been reduced.

For example, if we have several operations with registers
```
x++;
y = y + x;
z = 2 * y;
```

right now we reduce after every operation. you can have a flag per object + base
dom of whether they have been modified since the last reduction.

When you reach a reduction point, do not transfer from objects that have not been modified since the last reduction.

Also, if the operations only involve register arithmetic, there is nothing to reduce until there is a store using those registers.